README.md
## VPAID.js

Interact with methods and events controlling [VPAID 2.0](http://www.iab.net/media/file/VPAID_2.0_Final_04-10-2012.pdf)
video ad placements in pure Javascript. Manage and monitor your web video ads directly in-page,
rather than relying upon an overbuilt Flash video player. The goal of this framework is to provide
as near to IAB compliance as appropriate for external use.
 
### Installation

Grunt is a great tool for gathering build dependencies and putting together a predictable
environment for your builds. The pre-packaged environment from [grunt-air-sdk](https://www.npmjs.com/package/grunt-air-sdk)
makes for a very quick and easy build.

Gather all the dependencies

```
npm install
```

Build vpaidjs.swf using Flex 

```
grunt build
```

### Supported VPAID Events

  * AdLoading
  * AdLoaded
  * AdStarted
  * AdPaused
  * AdStopped
  * AdLinearChange
  * AdExpandedChange
  * AdVolumeChange
  * AdImpression
  * AdVideoStart
  * AdVideoFirstQuartile
  * AdVideoMidpoint
  * AdVideoThirdQuartile
  * AdVideoComplete
  * AdClickThru
  * AdUserAcceptInvitation
  * AdUserMinimize
  * AdUserClose
  * AdPlaying
  * AdLog
  * AdError
  * AdSkipped
  * AdSkippableStateChange
  * AdSizeChange
  * AdDurationChange
  * AdInteraction

Also, the ```AdReady```  event will be triggered when the ad placement has been loaded and is ready to start.

### Supported VPAID methods

  * initAd()
  * startAd()
  * resizeAd()
  * stopAd()
  * pauseAd()
  * resumeAd()
  * skipAd()
  * expandAd()
  * collapseAd()
  * volume()


CHANGELOG.md
1.0.0
-------------

* complete removal of [ova.as3](https://github.com/dave-hansen/ova.as3) library
* replaces ova with [AS3-IABLib](https://github.com/nathanhinish/AS3-IABLib) for VAST parsing and bootstrap
* VAST-only linear video support
  - may not be fully compliant


js/vpaid.latest.js
// version 0.1.1

var vpaidjs = vpaidjs || {};

// defaults
vpaidjs.options = {
  volume: 0.8,
  swfPath: "vpaidjs.swf",
  debug: false
};

var VPAID = function(playerId, options) {
  var player = this;
  this.ad = new Object;

  this.playerId = playerId;
  this.options = vpaidjs.options;

  for (var option in options) {
    player.options[option] = options[option];
  }

  this.create = function() {
    var flashvars = {};
    var params = {
      wmode: "transparent",
      allowScriptAccess: "always",
      bgcolor: "#000000"
    };

    // TODO: do i need these?
    var attributes = {
      id: player.playerId,
      name: player.playerId
    };

    swfobject.embedSWF(
      player.options.swfPath,
      player.playerId,
      "100%",
      "100%",
      "10.5",   // XXX: not sure what's safe here
      "",
      flashvars,
      params,
      attributes,
      onCreate
    );
  };

  /*
   *  VPAID Protocol Methods
   */

  this.initAd = function(config) {

    if (player.options.debug) {
      config.debug = {
        "levels": "fatal, config, vast_template, vpaid, http_calls, playlist, api"
      };
    }

    player.ad.initAd(config);
  };

  this.startAd = function() {
    player.ad.startAd();
  };

  this.resizeAd = function(x, y) {
    player.ad.resizeAd(x, y);
  };

  this.stopAd = function() {
    player.ad.stopAd();
  };

  this.skipAd = function() {
    player.ad.skipAd();
  };

  this.pauseAd = function() {
    // not implemented
  };

  this.resumeAd = function() {
    // not implemented
  };

  this.expandAd = function() {
    // not implemented
  };

  this.collapseAd = function() {
    // not implemented
  };

  this.volume = function(level) {
    // TODO: no idea why setting needs to happen twice, and only from JS
    player.ad.volume(level);
    player.ad.volume(level);
  };

  this.destroy = function() {
    swfobject.removeSWF(player.playerId);
  };

  // ridiculously overzealous way of verifying SWF fully loaded,
  // including the readiness of initAd() itself
  function onCreate(e) {
    if (!e.success || !e.ref ) {
      vpaidjs.log("Failed to embed SWF.")
      return false;
    }

    // wait just a smidge for Flash to start
    var readyCheck = setInterval(function () {
      if (typeof e.ref.PercentLoaded !== "undefined" && e.ref.PercentLoaded()) {
        clearInterval(readyCheck);
        // timer to wait for swf object to fully load
        var loadCheck = setInterval(function () {
          vpaidjs.log(player.options.swfPath + " " + e.ref.PercentLoaded() + "% loaded." );
          if (e.ref.PercentLoaded() === 100) {
            player.ad = document.getElementById(player.playerId);

            if (typeof player.ad.initAd == "function") {
              clearInterval(loadCheck);

              if (typeof player.options.success == "function") {
                player.options.success();
              }
            }
          }
        }, 100);
      }
    }, 100);
  }

  // now start it up
  this.create();
};

vpaidjs.log = function(message) {
  if (vpaidjs.options.debug) {
    try {
      window.top.console.log("vpaidjs: " + message);
    } catch(e) {}
  }
};

// External function called from vpaidjs.swf
vpaidjs.AdLog = function(message) {
  vpaidjs.log("[AdLog] " + message)
};
var __vpaidjs__ = window.vpaidjs = vpaidjs;


src/js/vpaid.js
// version 0.2a

var vpaidjs = vpaidjs || {};

// defaults
vpaidjs.options = {
  volume: 0.8,
  swfPath: "vpaidjs.swf",
  autoplay: false,
  debug: false
};
vpaidjs.activeAds = {};

var VPAID = function(playerId, options) {
  var player = this;
  this.ad = new Object();

  this.playerId = playerId;
  this.options = vpaidjs.options;

  for (var option in options) {
    player.options[option] = options[option];
  }

  this.create = function() {
    var flashvars = {};
    var params = {
      wmode: "transparent",
      allowScriptAccess: "always",
      bgcolor: "#000000"
    };

    // TODO: do i need these?
    var attributes = {
      id: player.playerId,
      name: player.playerId
    };

    swfobject.embedSWF(player.options.swfPath, player.playerId, "100%", "100%", "10.5", "", flashvars, params, attributes, onCreate);
  };

  this.initAd = function (adTag) {
    player.ad.initAd(adTag, player.options.debug);
  };

  this.resizeAd = function (x, y) {
    player.ad.resizeAd(x, y);
  };

  this.startAd = function () {
    player.ad.startAd();
  };

  this.stopAd = function() {
    player.ad.stopAd();
  };

  this.skipAd = function() {
    player.ad.skipAd();
  };

  this.pauseAd = function() {
    player.ad.pauseAd();
  };

  this.resumeAd = function() {
    player.ad.resumeAd();
  };

  this.expandAd = function() {
    player.ad.expandAd();
  };

  this.collapseAd = function() {
    player.ad.collapseAd();
  };

  this.volume = function (level) {
    player.ad.volume(level);
  };

  this.destroy = function () {
    if (typeof player.ad !== "undefined") {
      player.ad.stopAd();
    }
    swfobject.removeSWF(player.playerId);
  };

  this.on = function(eventName, cb) {
    // also accept lists of events and space-delimited event strings
    var events = typeof eventName === "object" ? eventName : eventName.split(" ");

    for (i in events) {
      player.ad.addEventListener(events[i], cb);
    }
  };

  // utilities
  function onCreate(e) {
    if (!e.success || !e.ref) {
      vpaidjs.log("Failed to embed SWF.");
      return false;
    }
    waitForSwfObject(e);
  }

  function waitForSwfObject(e) {
    var swfWait = setInterval(function () {
      if (typeof e.ref.PercentLoaded !== "undefined" && e.ref.PercentLoaded()) {
        clearInterval(swfWait);
        waitForAdInterface(e);
      }
    }, 100);
  }

  function waitForAdInterface(e) {
    // timer to wait for swf object to fully load
    var adWait = setInterval(function() {
      vpaidjs.log(player.options.swfPath + " " + e.ref.PercentLoaded() + "% loaded.");
      if (e.ref.PercentLoaded() === 100) {
        player.ad = document.getElementById(player.playerId);

        if (typeof player.ad.initAd == "function") {
          clearInterval(adWait);
		
          // add to list of active players
          vpaidjs.activeAds[player.playerId] = player;

          startEvents();
        }
      }
    }, 100);
  }

  function startEvents() {
    if (player.options.tag) {
      player.initAd(player.options.tag);
    }

    if (player.options.autoplay) {
      player.on("AdReady", function(e, data) {
        player.startAd();
      });
    }

    player.on("AdStopped", function(e, data) {
      delete vpaidjs.activeAds[player.playerId];
    });

    if (typeof player.options.success === "function") {
      player.options.success();
    }
  }

  // now start it up
  this.create();
};

vpaidjs.log = function(message) {
  if (vpaidjs.options.debug) {
    try {
      window.top.console.log("vpaidjs: " + message);
    } catch(e) {}
  }
};

// TODO: document ExternalInterface bridge
vpaidjs.triggerEvent = function(objectId, eventType, dataObj) {
  vpaidjs.log("[vpaid.js] event: " + eventType);

  var targetPlayer = window.document.getElementById(objectId);
  var vpaidEvent = new CustomEvent(eventType, { detail: JSON.parse(dataObj) });

  targetPlayer.dispatchEvent(vpaidEvent);
};

vpaidjs.VPAIDEvents = ["AdReady", "AdLoading", "AdLoaded", "AdStarted", "AdPaused", "AdStopped", "AdLinearChange", "AdExpandedChange", "AdVolumeChange", "AdImpression", "AdVideoStart", "AdVideoFirstQuartile", "AdVideoMidpoint", "AdVideoThirdQuartile", "AdVideoComplete", "AdClickThru", "AdUserAcceptInvitation", "AdUserMinimize", "AdUserClose", "AdPlaying", "AdLog", "AdError", "AdSkipped", "AdSkippableStateChange", "AdSizeChange", "AdDurationChange", "AdInteraction"];

var __vpaidjs__ = window.vpaidjs = vpaidjs;
